doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+        sequence_numbers = [sequence_numbers]
+    sequence_numbers = cast(List[int], sequence_numbers)  # mypy clarity
+```
+
+If you are using cast often, however, it might be an indication that the code should be otherwise written or refactored.
+Note that `typing.cast` is only used by the type checker and does not affect code at runtime.
+
+### Use forward references when the type is not defined yet
+
+This is commonly encountered when using `@classmethod`. Because the type does not exist yet, when trying to use it in a
+type hint, python complains at runtime.
+
+```python
+class TreeHouse:
+    def __init__(self): ...
+
+    @classmethod
+    def build(cls) -> TreeHouse:
+        return cls()
+```
+
+```cmd
+    class TreeHouse:
+  File "testing_mypy.py", line 12, in TreeHouse
+    def build(cls) -> TreeHouse:
+NameError: name 'TreeHouse' is not defined
+```
+
+A forward reference can be used to fix the runtime issue. Simply place the type in quotations like "TreeHouse":
+
+```python
+class TreeHouse:
+    def __init__(self) -> None: ...
+
+    @classmethod
+    def build(cls) -> "TreeHouse":
+        return cls()
+```
+
+`mypy` recognizes this syntax and will understand the type as `TreeHouse`.
+
+Another situation where forward references can be useful is when a type hint is used before a type is defined in a file.
+Here, `TreeHouse` gets defiend after `Yurt`, but `Yurt` uses it in a type hint so it must be a forward reference:
+
+```python
+class Yurt:
+    def add(house: "TreeHouse") -> None: ...
+
+
+class TreeHouse:
+    def __init__(self) -> None: ...
+```
+
+> Note: While this issue is easily resolved by adding `from __future__ import annotations`, this is currently only supported in Python 3.7+.
+
+### Use TYPE_CHECKING to avoid circular imports
+
+`typing.TYPE_CHECKING` is a constant that is evaluates to `True` when mypy is running, but is otherwise `False` at
+runtime. A common instance of when you might need to use `TYPE_CHECKING` is when a circular import is created due to
+imports necessary for type annotations:
+
+a.py
+
+```python
+from b import Baz
+
+
+def foo(b: Baz) -> None:
+    b.hello()
+
+
+def bar() -> None:
+    print("bar")
+```
+
+b.py
+
+```python
+from a import bar
+
+
+class Baz:
+    def hello(self) -> None: ...
+
+```
+
+Here `a` imports from `b` and `b` imports from `a`, causing a circular import at runtime:
+
+`ImportError: cannot import name 'Baz' from partially initialized module 'b' (most likely due to a circular import)`
+
+Since `TYPE_CHECKING` is only `True` when running the type checker, we can add our imports under a TYPE_CHECKING block
+to fix this issue in a.py:
+
+a.py
+
+```python
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from b import Baz
+
+
+def foo(b: "Baz") -> None:
+    b.hello()
+
+
+def bar() -> None:
+    print("bar")
+```
+
+Note that we also had to change Baz to a forward reference since it will not be defined at runtime under
```

Comment:
Also use it to hide types that are costly to load at runtime and useful only for typing

Also, typings shoult not be under this, since types from typings are free to import